src/api.ts

```diff
@@ -95,23 +89,19 @@ export async function startContainer(commitHash: CommitHash) {
 		[],
 		process.stdout,
 		{
+			...config.build.containerCreateOptions,
+			ExposedPorts: {'3000/tcp': {}},
```

Comment:
these two should also be in the config

Reply:
I was having trouble with these because of our use of `freePort` - maybe you can help me understand the need for `freePort` and its function.

Is it true that every branch takes on its own port? Does the app need to specify a port of its own? Is `3000` special or incidental?

Reply:
I'll try to explain how the ports are configured:

1. Within the container, calypso is always run on port 3000. This is specified in the wp-calypso codebase. Note that it is totally ok for every container to run calypso on the same guest port, since they each have their own sets of ports and aren't shared.  
2. The host OS running the container needs to connect the internal container port to one of its real ports (not virtualized). Each of these needs to be put on a unique port since we can't attach multiple processed to write to the same one without special multiplexing magic.  This is where freeport comes in to find an available port to attach it to. 
3. The Host OS  needs to maintain a map of all running containers and which port they are hooked up to so that it can proxy requests appropriately

Reply:
is it safe to assume we only need one external port though? maybe what we need is to generate the docker port configs automatically and have the config file indicate public ports to open up…?